### ISSUE TITLE
MIME-Header encoding broken in 2.80

### ISSUE THREAD

**Issue by jacquesg** [dankogai/p5-encode#105]
The following program illustrates a breakage that was introduced in 2.80:

```
use strict;
use warnings;
use Encode qw(encode);

my $headers = join
(
        "\r\n",
                'From: koos@somewhere.com',
                'Custom-Header: 123',
                ''
);

print encode ('MIME-Header', $headers);
```

This produces the following for 2.88
```
=?UTF-8?B?RnJvbToga29vc0Bzb21ld2hlcmUuY29tDQpDdXN0b20tSGVhZGVyOiAxMjMN?=
 =?UTF-8?B?Cg==?=
```

The same program produce the following for 2.78:
```
From: koos@somewhere.com
Custom-Header: 123
```

**Comment by pali** [comment]
https://metacpan.org/pod/Encode::MIME::Header#BUGS

**Comment by jacquesg** [comment]
This does not address the fact that the header field names themselves are (unnecessarily) encoded. My interpretation of the RFCs is that encoded-words may be used for the ```field-body```. IMHO it does not make provision for encoding ```field-name```. See below from RFC 2047

> 
> 6. Support of 'encoded-word's by mail readers
> 
> 6.1. Recognition of 'encoded-word's in message headers
> 
>    A mail reader must parse the message and body part headers according
>    to the rules in RFC 822 to correctly recognize 'encoded-word's.
> 
>    'encoded-word's are to be recognized as follows:
> 
>    (1) Any message or body part header field defined as '*text', or any
>        user-defined header field, should be parsed as follows: **Beginning
>        at the start of the field-body and immediately following each
>        occurrence of 'linear-white-space', each sequence of up to 75
>        printable characters (not containing any 'linear-white-space')
>        should be examined to see if it is an 'encoded-word' according to
>        the syntax rules in section 2.  Any other sequence of printable
>        characters should be treated as ordinary ASCII text.**
>

**Comment by pali** [comment]
Hi! The whole problem is much more complicates as you see. I know meaning of those RFCs as I worked on fixing MIME encoding/decoding in perl for about 2 years.

Reason for change was finally fix numerous bugs in MIME encoder and decoder as written in BUGS section. Basically old implementation was incorrect and produced invalid output which must not be sent to world. Couple of software, specially those which were written correct according to RFCs, was unable to process emails encoded by this module. You can search for bugs in rt tracker for Encode module.

The whole problem is in fact that without semantic knowledge it is not possible to encode/decode whole email structure (like stored in mbox file). Such thing is MIME-Header not doing, nor is suppose to do. Previous version tried to encode/decode something unspecified which I do not know what it was. But output was for any my interpretation incorrect and not compliant to RFC2822 nor RFC2047.

MIME-Header is there for doing encoding and decoding of unstructured field body of the email header. Or for 'text' token. If you need anything else, like encoding/decoding whole email body or whole email (in mbox format), you need to use module which understand whole email structure as Email::MIME.

Please look at documentation https://metacpan.org/pod/Encode::MIME::Header where is written what this module is doing. It does not guaranteed that module produce same output for different version, but now I would like to say that it produce output compliant to RFC2047 and that it contains correct representation of input data. For one input unicode string it is possible to produce more different MIME-encoded outputs, but their decoded representation would be same as input.

**Comment by jacquesg** [comment]
The contract of the module was changed in the documentation between version ``2.86`` and ``2.87``.

This is really not nice considering that this is a **core module**.

**Comment by pali** [comment]
Old implementation was broken, produced invalid output and was not suitable for usage, specially for communication with outside world. Its design was broken and not possible to fix. People asked for fixing it and there started to exist some modules which doing similar thing or reimplement some of functionality. Version 2.83 was released more then year ago and you are probably the first one complaining about this change.

Sending incorrect, broken and non-compliant emails to word is something which must be eliminated. Plus it show bad light on perl itself which produce such garbage. There was basically only two options: Fix module or completely remove it. Both cases would lead to problem which you described.

If you want to generate/produce MIME encoded strings do not use Encode prior to 2.83 as it does not work and produce incorrect output.

**Comment by jacquesg** [comment]
I think you are missing the point. Existing code that relied on encoding the whole MIME-Header now creates even less compliant e-mails! This was a silent deliberate breakage which is not acceptable. The fact that this got reported 1 year later is irrelevant. Not everyone lives on the bleading edge.

The changes that were made to this module is not in line with perl's own polices in terms of backward-compatibility and deprecation:
http://perldoc.perl.org/perlpolicy.html#BACKWARD-COMPATIBILITY-AND-DEPRECATION

> Historically, we've held ourselves to a far higher standard than backward-compatibility -- bugward-compatibility. Any accident of implementation or unintentional side-effect of running some bit of code has been considered to be a feature of the language to be defended with the same zeal as any other feature or functionality. **No matter how frustrating these unintentional features may be to us as we continue to improve Perl, these unintentional features often deserve our protection. It is very important that existing software written in Perl continue to work correctly. If end-user developers have adopted a bug as a feature, we need to treat it as such**.

**Comment by pali** [comment]
I'm not missing the point. It does not worked at all, it was undocumented and produced something not suitable for emails. Now it has predicable and fully documented behavior. There was already decision to fix it and Encode maintainer accepted it.

Bugs are fixing in time and of course every fix slightly change algorithms. You can downgrade back to Perl 5.6 and you would have perl again with bugs, but with same buggy behavior.

If you do not agree then open discussion with maintainers. As I do not understand from your comments what are you trying to achieve now.

**Comment by jacquesg** [comment]
I do not agree that it was undocumented, I would rather say that it was less specific or underspecified.

Did it always produce something that was suitable for e-mails? I do not believe so. I'm providing another example:

```
use strict;
use warnings;
use utf8;
use Encode qw(encode);

my $headers = join
(
        "\r\n",
                'From: voël@services.test.lan',
                'To: jacquesg@services.test.lan',
                'Subject: Encode test',
                'Custom-Header: 123',
);

print encode ('MIME-Header', $headers)."\r\n\r\nThis is an encode test";
```

This produces in 2.78:

```
From: =?UTF-8?B?dm/Dq2xAc2VydmljZXMudGVzdC5sYW4=?=
To: =?UTF-8?B?amFjcXVlc2dAc2VydmljZXMudGVzdC5sYW4=?=
Subject: =?UTF-8?B?RW5jb2RlIHRlc3Q=?=
Custom-Header: =?UTF-8?B?MTIz?=

This is an encode test
```

In 2.89 it gives:
```
=?UTF-8?B?RnJvbTogdm/Dq2xAc2VydmljZXMudGVzdC5sYW4NClRvOiBqYWNxdWVzZ0Bz?=
 =?UTF-8?B?ZXJ2aWNlcy50ZXN0Lmxhbg0KU3ViamVjdDogRW5jb2RlIHRlc3QNCkN1c3Rv?=
 =?UTF-8?B?bS1IZWFkZXI6IDEyMw==?=

This is an encode test
```

The version produced by 2.78 is perfectly legal whereas 2.89 is not.

**Comment by jacquesg** [comment]
@dankogai I believe you are the maintainer of this distribution, can you maybe give your input?

**Comment by pali** [comment]
> This produces in 2.78:
> ```
> From: =?UTF-8?B?dm/Dq2xAc2VydmljZXMudGVzdC5sYW4=?=
> To: =?UTF-8?B?amFjcXVlc2dAc2VydmljZXMudGVzdC5sYW4=?=
> ...
> ```
> ...
> The version produced by 2.78 is perfectly legal

It is not legal and is incorrect too.

**Comment by jacquesg** [comment]
Please elaborate why this is incorrect?

**Comment by jacquesg** [comment]
Here is a script written in python by somebody else
https://github.com/akkana/scripts/blob/master/decodemail.py

```
jacquesg@jacquesg:~/tmp clang-64 $ cat msg.txt
From: =?UTF-8?B?dm/Dq2xAc2VydmljZXMudGVzdC5sYW4=?=
To: =?UTF-8?B?amFjcXVlc2dAc2VydmljZXMudGVzdC5sYW4=?=
Subject: =?UTF-8?B?RW5jb2RlIHRlc3Q=?=
Custom-Header: =?UTF-8?B?MTIz?=

This is an encode test
jacquesg@jacquesg:~/tmp clang-64 $ python decodemail.py From msg.txt
From: voël@services.test.lan
```

**Comment by jacquesg** [comment]
Also try this one http://dogmamix.com/MimeHeadersDecoder/

**Comment by pali** [comment]
> Please elaborate why this is incorrect?

Yes. It violates grammar described in RFC2822 section 3.6.2 for From header. Same for To header described in 3.6.3. From And To headers are structured and must contain mailbox-list resp. address-list. See section 3.4 for specification. It is **not** one single word, it is some structure which has addr-spec.

**Comment by jacquesg** [comment]
Sure, I can concede that it is not 100% according to the RFC. In practice mails generated with previous version had a better chance of being delivered. Fact remains that there is no perldelta entry for the behaviour change.

**Comment by pali** [comment]
So you agreed it was not correct too. And this is reason why is was fixed. There are many other systems which were not able to process such garbage. (one example pipermail, part of mailman, written in python).

@dankogai Any comments? Or can be this issue now closed?

If there were missing entires in perldelta or other parts which of perl itself, please report this problem to perl (via perlbug). Encode cpan module on cpan does not maintain any own perldelta, there is just Changes file filled by @dankogai at cpan upload time.

**Comment by jacquesg** [comment]
I still maintain that "bugward" compatibility as per the core perl rules should have remained.

**Comment by jacquesg** [comment]
See here: https://perldoc.perl.org/perlpolicy.html#BACKWARD-COMPATIBILITY-AND-DEPRECATION

**Comment by jacquesg** [comment]
> Historically, we've held ourselves to a far higher standard than backward-compatibility -- bugward-compatibility. Any accident of implementation or unintentional side-effect of running some bit of code has been considered to be a feature of the language to be defended with the same zeal as any other feature or functionality. No matter how frustrating these unintentional features may be to us as we continue to improve Perl, these unintentional features often deserve our protection. It is very important that existing software written in Perl continue to work correctly. If end-user developers have adopted a bug as a feature, we need to treat it as such.

**Comment by pali** [comment]
That applies for perl itself, not for new modules on cpan.

And this situation with MIME is slightly different. Here we are dealing with interoperability with other systems and it was purely a bug which made perl unusable for such MIME-encoding action. There were a lot of bug reports and complains of users of other systems that Encode perl module was sending garbage to the world, which was not compliant and not parsable by other systems. Older behavior just caused problems for all people and lot of them were not aware of the fact that they are sending garbage in their emails to the world. Such garbage is not acceptable in the world and it is better to stop people generating or sending such emails. If it throws error, then it is better as silently do something undefined and let receiver or email server start screeming.

Above backward compatibility make sense for between perl and perl, but not for something like world-wide standard which perl generates and expects that whole non-perl world implement perl's own modification.

Also, see how the whole problem was handled. Encoder is RFC-strict, but decoder is not. Decoder is still trying to decode strings generated by the older version of Encode module, just for backward compatibility and ability to process emails which were produced by older Encode versions. This can be enabled/disabled by the `$Encode::MIME::Header::STRICT_DECODE` variable, and by default is **enabled**. Also note that there is a huge test coverage and specially also cases for data generated by the older Encode version -- just to ensure decode ability.

So basically backward compatibility for Perl is still there, which ensures that old "data" are still parseable by Encode module and perl itself.

This is basically the thing which could be done. Users start seeing warning and errors when something wrong want to send to the world and users still can receive such garbage from the world (if they explicitly do not disable it).

**Comment by ernix** [comment]
@pali Please, please keep the original behaviour.

I know that the MIME is not so simple, old implementation is quirky and Encode modules has no responsible to parse/produce RFC2047 address syntax.  BUT the new implementation breaks every single scripts rely on old fashioned `sendmail(1)` and `Encode::MIME::Header` silently.

That's exactly why `Encode::MIME::Header` is listed in core module, isn't it?
New implementation does completely different things and lacks consistency between `Encode::MIME::Header` and `Encode::MIME::Header::ISO_2022_JP`.

Could you give me an alternate solution or migration path, please?  I really need your help.
`Email::MIME` is not core module, `Email::MIME::RFC2047` requires perl5.10, I want something like `$Encode::MIME::Header::STRICT_ENCODE` to make them work.

**Comment by pali** [comment]
> Could you give me an alternate solution or migration path, please? I really need your help.

What exactly do you need? Or what are you trying to achieve?

> ... requires perl5.10

And under which perl version?

As already stated old Encode::MIME::Header generated totally invalid strings which are not suitable for including into MIME emails. Therefore using old Encode::MIME::Header for generating emails is a wrong idea.

**Comment by ernix** [comment]
@pali Following is a sample code.  It's running under perl 5.8.8.

```perl
#!/usr/bin/perl
use strict;
use warnings;
use utf8;
use Encode;

my $from = '"児島 新" <shin@kojima.org>';
my $to = '"児島 新" <shin@.kojima.org>';
my $subj = 'こんにちわーるど';

my $msg = <<__MAIL;
${\( encode MIME_Header => "From: $from" )}
${\( encode MIME_Header => "To: $to" )}
${\( encode MIME_Header => "Subject: $subj" )}
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Hello world
.
__MAIL

open my $fh, '|-', qw{/usr/sbin/sendmail -t -n -oi};
for my $line (split /\r?\n/, $msg) {
    print encode_utf8 "$line\n";
}
close $fh;

1;
```

**Comment by pali** [comment]
So you are trying to encode arbitrary email header, even structured like To? Encode::MIME::Header never worked on structured headers (as it needs to know full RFC2822 grammar of structured header) and if tried then it generated garbage.

The easiest option is to use Email::Stuffer module at least in version 0.016. That module uses Email::MIME module at least in version 1.943 which is responsible for setting all those From/To/... structured headers correctly (via header_str_set resp. header_str).

Example:
```
use Email::Stuffer 0.016;
use utf8;

my $from = '"児島 新" <shin@kojima.org>';
my $to = '"児島 新" <shin@.kojima.org>';
my $subj = 'こんにちわーるど';
my $body = <<__MAIL;
Hello world
__MAIL

my $msg = Email::Stuffer->from($from)
                        ->to($to)
                        ->subject($subject)
                        ->text_body($body)
                        ->as_string();
```

Email::Stuffer has also send method.

There is not and never was core module for parsing structured email headers.

**Comment by dankogai** [comment]
Since I agree with @pali 100%, I have nothing further to say.  Case dismissed.
